pumpkin-config/src/resource_pack.rs

```diff
@@ -25,6 +27,21 @@ impl ResourcePackConfig {
             "Resource pack sha1 hash is too long (max. 40)"
         )
     }
+
+    pub fn get_pack(&self) -> CConfigAddResourcePack<'_> {
+        let prompt_message: Option<TextComponent<'_>> = if self.prompt_message.is_empty() {
+            None
+        } else {
+            Some(TextComponent::text(&self.prompt_message))
+        };
+        CConfigAddResourcePack::new(
+            uuid::UUID::new(self.resource_pack_url.as_bytes()),
+            &self.resource_pack_url,
+            &self.resource_pack_sha1,
+            self.force,
+            prompt_message,
+        )
+    }
```

Comment:
Pumpkin Config should not have to use any Packets at all

Reply:
The alternative is a new type wrapper around the config in the `pumpkin` crate. While I agree there should be strict dependency rules around crates, (after all I was the one who proposed making such rules), this is the location which provides exactly enough information to produce this packet. Refactoring either the packet itself or the resource pack data would require changes from both, so coupling the two together means you only have to change two locations instead of three.

If you actually want me to create a newtype for this I can.

Reply:
> The alternative is a new type wrapper around the config in the `pumpkin` crate. While I agree there should be strict dependency rules around crates, (after all I was the one who proposed making such rules), this is the location which provides exactly enough information to produce this packet. Refactoring either the packet itself or the resource pack data would require changes from both, so coupling the two together means you only have to change two locations instead of three.
> 
> If you actually want me to create a newtype for this I can.

you can drop the `pumpkin-protocol` dep, then add `pumpkin-config` in protocol, then you can impl `AsRef<CConfigAddResourcePack>` or something similar, maybe there is a better way to do this though.

Reply:
I would tend to believe that's an even worse structuring to have the low level protocol API dependent on the high level server configuration.

In terms of layers of abstraction, it should be

Protocol -> Config -> pumpkin

Having 

Config -> Protocol -> pumpkin

Puts config much further down the dependency tree.

Why do you believe it's more appropriate to have the config class as a protocol dependency?

Ostensibly, the protocol will change little, if any, and only additively. Using it as a foundational crate for everything else makes significantly more sense than the other way around.

Reply:
> Protocol -> Config -> pumpkin
> 
> Having
> 
> Config -> Protocol -> pumpkin
> 
> Puts config much further down the dependency tree.

I would argue that having config down the chain is better, protocol will most likely depend on config anyway, config will also contain lower-level configuration from features.toml, which alex wants to be able to configure **everything** with.

> Why do you believe it's more appropriate to have the config class as a protocol dependency?

Its because protocol depends on a lot of things, which would mean that if config depends on protocol, those things would not be able to depend on config.

> Ostensibly, the protocol will change little, if any, and only additively. Using it as a foundational crate for everything else makes significantly more sense than the other way around.

Thats true in theory, but for packets where serialization is more complicated, when we change the internal data structure, we will also need to update the serialization.

---

Imo config should only ever depend on core, otherwise it will be annoying to use it, and pumpkin-protocol should only be a dep in the main crate, this is not because of some rule, but due to how everything is currently structured.

If you think theres a better solution then im open to suggestions!

PS: please dont just suggest switching all things around (from pumpkin-protocol depending on everything to everything depending on pumpkin-protocol), that is not a better solution, just a different one, and it also makes it impossible to compose packets using types from other crates in the packets

Reply:
> Its because protocol depends on a lot of things, which would mean that if config depends on protocol, those things would not be able to depend on config.

This is just not true. The protocol is defined by Mojang and the Minecraft API. There is nothing to configure. These are fixed types that are not configurable. You can build a packet or message based on what you may care about, but the inputs and serialization are not configured.

Put another way, there is only one way to write a packet to a Minecraft client, and there is only one way a Minecraft server will receive that packet from the client. A packet's shape, or layout will never, and can never change based on server configuration. The *only* way you can configure a server is to decide what packets you will send or not.

Serialization doesn't change based on configuration. You just change *what* you serialize based on configuration. Something which belongs outside of the `protocol` crate, because it's not a part of the protocol itself.

Everything *but* protocol types are configurable here.

It might be good for you to understand what "ostensibly" means.

Reply:
> This is just not true. The protocol is defined by Mojang and the Minecraft API. There is nothing to configure. These are fixed types that are not configurable. You can build a packet or message based on what you may care about, but the inputs and serialization are not configured.
> 
> Put another way, there is only one way to write a packet to a Minecraft client, and there is only one way a Minecraft server will receive that packet from the client. A packet's shape, or layout will never, and can never change based on server configuration. The only way you can configure a server is to decide what packets you will send or not.

but pumpkin-protocol does include serialization, it was agreed upon this way, so that serialization isnt placed around in random places around the codebase.
also, some packets cant be serialized from their serde layout, so using only types that are described in wiki.vg is impossible, idk if youve ever read the wiki.vg, but its filled with "which varies by context", "depends on context", "must be known from the context"

> Serialization doesn't change based on configuration. You just change *what* you serialize based on configuration. Something which belongs outside of the protocol crate, because it's not a part of the protocol itself.

alr, but pumpkin-protocol depends on other crates that will depend on config so you cant do it anyway

> It might be good for you to understand what "ostensibly" means.

Communication is a skill in and of itself, using English words that are archaic or uncommon isn't adroit in most situations, so using a *fancy* word just to use them is actually very counterproductive. 
also fyi i know what it means, if you had actually read what i wrote comprehensively, youd know that i disagreed with what you have said

Reply:
> but pumpkin-protocol does include serialization, it was agreed upon this way, so that serialization isnt placed around in random places around the codebase.
also, some packets cant be serialized from their serde layout, so using only types that are described in wiki.vg is impossible, idk if youve ever read the wiki.vg, but its filled with "which varies by context", "depends on context", "must be known from the context"

Nothing said here is contradicting anything I said. I agree that pumpkin-protocol should handle serialization of packet types. That's what a protocol is. It's absolutely true that some packets are context dependent, and cannot be serialized without knowing, for example, connection state. But nothing said here argues otherwise.